### PR TITLE
Add PlainText Function

### PR DESCRIPTION
**Pull request by nguhoa**
Fix issue: https://github.com/microsoft/Power-Fx/issues/1429

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 1]
```diff
@@ -1059,6 +1068,34 @@ public static FormulaValue OptionSetValueToLogicalName(IRContext irContext, Opti
             var optionSet = args[0];
             var logicalName = optionSet.Option;
             return new StringValue(irContext, logicalName);
+        }
+
+        public static FormulaValue PlainText(IRContext irContext, StringValue[] args)
```
Comment (anderson-joyle) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:1073]: >public static FormulaValue PlainText(IRContext irContext, StringValue[] args) [](http://example.com/codeflow?start=8&length=77)

How `PlainText` function should handle a blank value? PA returns `Blank()` while this implementation is returning `Error({Kind:ErrorKind.InvalidArgument})`

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs [hunk 2]
```diff
@@ -1705,10 +1705,21 @@ static Library()
             {
                 BuiltinFunctionsCore.Language,
                 NoErrorHandling(Language)
-            }
+            },
+            {
+                BuiltinFunctionsCore.PlainText,
```
[the remark this answers is not in the document]
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs]: ordered them in alphabet. thanks

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 1 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs]: Checked on PowerApps side and updated logic to allow formularvalue input to fix here. Thanks.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        public PlainTextFunction()
+            : base("PlainText", TexlStrings.AboutPlainText, FunctionCategories.Text, DType.String, 0, 1, 1, DType.String)
+        {
+        }
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures()
+        {
+            yield return new[] { TexlStrings.PlainTextArg1 };
+        }
+    }
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs:27]: Can we move already existing Implementation from PA?
(Luc recently moved regex function from PA so may have pointers on how to do it w/o breaking PA)
[https://msazure.visualstudio.com/OneAgile/_git/PowerApps-Client?path=/src/Cloud/DocumentServer.Core/Document/Texl/Builtins/PlainText.cs&version=GBmaster](https://msazure.visualstudio.com/OneAgile/_git/PowerApps-Client?path=/src/Cloud/DocumentServer.Core/Document/Texl/Builtins/PlainText.cs&version=GBmaster)

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs [hunk 4]
```diff
@@ -939,6 +939,10 @@ static Library()
                     returnBehavior: ReturnBehavior.ReturnFalseIfAnyArgIsBlank,
                     targetFunction: IsToday)
             },
+            {
+                BuiltinFunctionsCore.Language,
+                NoErrorHandling(Language)
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:944]: >NoErrorHandling [](http://example.com/codeflow?start=16&length=15)

Why No Error Handling?

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 5]
```diff
@@ -1059,6 +1069,39 @@ public static FormulaValue OptionSetValueToLogicalName(IRContext irContext, Opti
             var optionSet = args[0];
             var logicalName = optionSet.Option;
             return new StringValue(irContext, logicalName);
+        }
+
+        public static FormulaValue PlainText(IRContext irContext, FormulaValue[] args)
+        {
+            if (args[0].Type == FormulaType.String)
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:1076]: >if (args[0].Type == FormulaType.String) [](http://example.com/codeflow?start=12&length=39)

Can we avoid this check and use error handling pipeline maybe?

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 4 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:944]: It has `returnBehavior: ReturnBehavior.ReturnBlankIfAnyArgIsBlank` which may be useful ;)

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+"-12.34"
+
+>> PlainText(Date(2023,1,1))
+"1/1/2023"
+
+>> PlainText(DateTime(2023,1,1,12,34,56))
+"1/1/2023 12:34 PM"
+
+>> PlainText(Time(8,1,15))
+"8:01 AM"
```
Comment (jas-valgotar) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt:62]: Can you add negative tests if it applies? Ie. where the function would return an error.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 4 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:944]: This is language function, not in scope of this PR. but this is what Calos suggested.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs (hunk 3 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs:27]: not sure what this mean but this function is trying to implement the same way with PA.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 5 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:1076]: this is not error handling. If it is string format then I tried to decode it.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 4 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:944]: Oops my bad thought it was PlainText :D

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs [hunk 7]
```diff
@@ -1204,6 +1208,17 @@ static Library()
...
             },
+            {
+                BuiltinFunctionsCore.PlainText,
+                StandardErrorHandling<FormulaValue>(
+                    BuiltinFunctionsCore.PlainText.Name,
+                    expandArguments: NoArgExpansion,
+                    replaceBlankValues: NoOpAlreadyHandledByIR,
+                    checkRuntimeTypes: DeferRuntimeTypeChecking,
+                    checkRuntimeValues: DeferRuntimeValueChecking,
+                    returnBehavior: ReturnBehavior.AlwaysEvaluateAndReturnResult,
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1219]: >ReturnBehavior.AlwaysEvaluateAndReturnResult [](http://example.com/codeflow?start=36&length=44)

maybe `ReturnBehavior.ReturnBlankIfAnyArgIsBlank ` ?

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs [hunk 8]
```diff
@@ -1204,6 +1208,17 @@ static Library()
                 BuiltinFunctionsCore.Pi,
                 NoErrorHandling(Pi)
             },
+            {
+                BuiltinFunctionsCore.PlainText,
+                StandardErrorHandling<FormulaValue>(
+                    BuiltinFunctionsCore.PlainText.Name,
+                    expandArguments: NoArgExpansion,
+                    replaceBlankValues: NoOpAlreadyHandledByIR,
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1216]: >NoOpAlreadyHandledByIR [](http://example.com/codeflow?start=40&length=22)

Should this be `DoNotReplaceBlank` ?

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs [hunk 9]
```diff
@@ -1204,6 +1208,17 @@ static Library()
...
+                    checkRuntimeTypes: DeferRuntimeTypeChecking,
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1217]: >DeferRuntimeTypeChecking [](http://example.com/codeflow?start=39&length=24)

Should this be `ExactValueTypeOrBlank<StringValue>` ?

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 5 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:1076]: When would this be not string type? (Wouldn't Checktype already catch it?)

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 7 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1219]: Looks like ReturnBlankIfAnyArgIsBlank has been used for functions will multiple args and make sure return blank if any of them blank. But this plaintext function has only 1 input. and logic of it return blank if input blank already.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs (hunk 3 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs:27]: Ideally, we want PA and FX to share a TexlFunction file.
Currently this file already exists in PA repo, so Ideally we should remove it from PA repo and move it to here FX repo...

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 8 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1216]: I think this is fine. input is blank then output blank. so default let IR handle

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 9 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1217]: No. want the value is formulavalue.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt (hunk 6 above)
Comment (nguhoa) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt:62]: Yes, added more error test here.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+>> PlainText("  <  xml><p>Input<tr>  </xml>")
+"Input"
+
+>> PlainText("http&#x3a;//microsoft.com")
+"http://microsoft.com"
+
+>> PlainText("http&#58;//microsoft.com")
+"http://microsoft.com"
+
+>> PlainText(Blank())
```
Comment (jas-valgotar) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt:46]: >PlainText(Blank()) [](http://example.com/codeflow?start=3&length=18)

Can you please also add typed blank in the test
e.g. `PlainText(If(1<0, "text"))`

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 8 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1216]: I think we set this to `NoOpAlreadyHandledByIR` when the IR actually converts Blank to something. Otherwise `DoNotReplaceBlank` is preferred.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 9 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1217]: I think Line 1213 Also needs to be `StringValue`

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 7 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1219]: Correct, I think the same should work for 1 Arg as well.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt (hunk 10 above)
Comment (nguhoa) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt:46]: Added this to test.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+Blank()
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt:47]: Blank values, when passed to arguments that expect strings, should be interpreted as empty strings. So this should return `""`

Comment (anderson-joyle) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt:47]: Thanks for correcting me on this. `IsBlank(PlainText(Blank()))` returns true in PA, which made me think `Blank()` would be the right result. But then I realized that `IsBlank("")` is also true in PA.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs (hunk 3 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs:27]: it is the same content file. what do you want to move from PA to Fx?

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 5 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:1076]: i just allow they input differnt format here: string, blank() or datetime or boolean. but the decoding only process for string.

File: src/strings/PowerFxResources.en-US.resx [hunk 12]
```diff
@@ -5158,4 +5158,15 @@
     <value>Can't delegate {0}: contains a behavior function '{1}'.</value>
     <comment>Warning message.</comment>
   </data>
+  <data name="AboutPlainText" xml:space="preserve">
+    <value>The PlainText function removes HTML and XML tags.</value>
+    <comment>Description of 'PlainText' function.</comment>
+  </data>
+  <data name="PlainTextArg1" xml:space="preserve">
+    <value>PlainText</value>
```
Comment (CarlosFigueiraMSFT) [inline src/strings/PowerFxResources.en-US.resx:5166]: >PlainText [](http://example.com/codeflow?start=11&length=9)

`text` (this is the text that is going to be "plainified")

File: src/strings/PowerFxResources.en-US.resx [hunk 13]
```diff
@@ -5158,4 +5158,15 @@
...
+    <comment>Function_parameter - First parameter for the PlainText function</comment>
+  </data>
+  <data name="AboutPlainText_PlainText" xml:space="preserve">
```
Comment (CarlosFigueiraMSFT) [inline src/strings/PowerFxResources.en-US.resx:5169]: >PlainText [](http://example.com/codeflow?start=29&length=9)

`text`

File: src/strings/PowerFxResources.en-US.resx [hunk 14]
```diff
@@ -5158,4 +5158,15 @@
     <value>Can't delegate {0}: contains a behavior function '{1}'.</value>
     <comment>Warning message.</comment>
   </data>
+  <data name="AboutPlainText" xml:space="preserve">
+    <value>The PlainText function removes HTML and XML tags.</value>
```
Comment (CarlosFigueiraMSFT) [inline src/strings/PowerFxResources.en-US.resx:5162]: >The PlainText function removes HTML and XML tags. [](http://example.com/codeflow?start=11&length=49)

We should keep the existing description (unless it is buggy): "Removes all formatting and returns a plain text value."

File: src/strings/PowerFxResources.en-US.resx [hunk 15]
```diff
@@ -5158,4 +5158,15 @@
...
+    <comment>Description of 'PlainText' function.</comment>
+  </data>
+  <data name="PlainTextArg1" xml:space="preserve">
+    <value>PlainText</value>
+    <comment>Function_parameter - First parameter for the PlainText function</comment>
```
Comment (CarlosFigueiraMSFT) [inline src/strings/PowerFxResources.en-US.resx:5167]: >Function_parameter - First parameter for the PlainText function [](http://example.com/codeflow?start=13&length=63)

"function_parameter - First argument of the PlainText function - the text to have its HTML and XML tags removed."

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 8 above)
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1216]: >NoOpAlreadyHandledByIR [](http://example.com/codeflow?start=40&length=22)

If we say that it's handled by the IR, we probably need to override GetArgPreprocessor in the function (or make it inherit from StringOneArgFunction instead of BuiltinFunction)

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 16]
```diff
@@ -41,6 +43,14 @@ internal static partial class Library
         private static readonly Regex _minutesDetokenizeRegex = new Regex("[\u000A][\u000A]+", RegExFlags);
         private static readonly Regex _secondsDetokenizeRegex = new Regex("[\u0008][\u0008]+", RegExFlags);
         private static readonly Regex _milisecondsDetokenizeRegex = new Regex("[\u000e]+", RegExFlags);
+        private static readonly Regex _tdTagRegex = new Regex("<\\s*([tT][dD])\\s*\\/{0,1}>", RegExFlags);
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:46]: >RegExFlags [](http://example.com/codeflow?start=94&length=10)

Instead of having the casing done in the regex string we can use RegexOptions.IgnoreCase in those regexes

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 5 above)
Comment (CarlosFigueiraMSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:1076]: Since this function declares its argument as string, it will only receive a string - if we pass a different data type it will be first coerced to a string prior to being passed to the function. You can check on the REPL project to see what happens under the hood.

For example, if you enter `IR(Len(Today()))` in the repl console, you'll see the call to the Text function (which implements Date[Time]-to-text coercion) in the IR:

    Len:w(
        BlankToEmptyString:s(
            Text:s(Today:D(), "'shortdate'":s)))

Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:1076]: I see. thanks for explaining. i changed this back to stringvalue set.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 8 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1216]: Ah I see, I updated this to replace empty string for blank value,

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 16 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:46]: I am trying to use the same pattern with other Regex where we want to use RegExFlags. should I use IgnoreCase here instead?

File: src/strings/PowerFxResources.en-US.resx (hunk 15 above)
Comment (nguhoa) [reply src/strings/PowerFxResources.en-US.resx:5167]: Update here

File: src/strings/PowerFxResources.en-US.resx (hunk 14 above)
Comment (nguhoa) [reply src/strings/PowerFxResources.en-US.resx:5162]: Updated. thanks

File: src/strings/PowerFxResources.en-US.resx (hunk 12 above)
Comment (nguhoa) [reply src/strings/PowerFxResources.en-US.resx:5166]: update

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt (hunk 11 above)
Comment (nguhoa) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt:47]: Updated logic and now it return ""

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 16 above)
Comment (CarlosFigueiraMSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:46]: I believe we can move all of them to use IgnoreCase - it will make it much easier to read

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 17]
```diff
@@ -1059,6 +1069,34 @@ public static FormulaValue OptionSetValueToLogicalName(IRContext irContext, Opti
...
+            text = _styleTagRegex.Replace(text, string.Empty);
+
+            // Remove all comments.
+            text = _commentTagRegex.Replace(text, string.Empty);
+
+            // Insert empty string in place of <td>
+            text = _tdTagRegex.Replace(text, string.Empty);
+
+            //Replace <br> or <li> with line break
+            text = _lineBreakTagRegex.Replace(text, Environment.NewLine);
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:1091]: >text = _lineBreakTagRegex.Replace(text, Environment.NewLine); [](http://example.com/codeflow?start=12&length=61)

We should have tests for this case. And for the next one.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 16 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs]: updated here.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 17 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:1091]: added tests.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 18]
```diff
@@ -20,18 +22,18 @@ namespace Microsoft.PowerFx.Functions
...
     internal static partial class Library
     {
         private static readonly RegexOptions RegExFlags = LibraryFlags.RegExFlags;
 
-        private static readonly Regex _ampmReplaceRegex = new Regex("[aA][mM]\\/[pP][mM]", RegExFlags);
-        private static readonly Regex _apReplaceRegex = new Regex("[aA]\\/[pP]", RegExFlags);
-        private static readonly Regex _minutesBeforeSecondsRegex = new Regex("[mM][^dDyYhH]+[sS]", RegExFlags);
-        private static readonly Regex _minutesAfterHoursRegex = new Regex("[hH][^dDyYmM]+[mM]", RegExFlags);
-        private static readonly Regex _minutesRegex = new Regex("[mM]", RegExFlags);
+        private static readonly Regex _ampmReplaceRegex = new Regex("am\\/pm", RegExFlags);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:32]: why the change in regexs?

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/PlainTextTests.cs [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+        [Theory]
+        [InlineData("PlainText(\"Test<br id = 'br1'>break line</  BR>\")", "Test\r\nbreak line")]
+        [InlineData("PlainText(\"1<br>2<li>3\")", "1\r\n2\r\n3")]
+        [InlineData("PlainText(\"1<BR/>2<li/>3\")", "1\r\n2\r\n3")]
+        [InlineData("PlainText(\"Test<p >double</  p><div >break line</ Div>\")", "Test\r\n\r\ndouble\r\n\r\nbreak line")]
+        [InlineData("PlainText(\"1<div>2<p>3<tr>4\")", "1\r\n\r\n2\r\n\r\n3\r\n\r\n4")]
+        [InlineData("PlainText(\"1<Div/>2<p/>3<Tr/>4\")", "1\r\n\r\n2\r\n\r\n3\r\n\r\n4")]
+        [InlineData("PlainText(\"1<  br  >2<  Li />3\")", "1\r\n2\r\n3")]
+        [InlineData("PlainText(\"1<div  class = 'div2'>2< P />3<  tr>4\")", "1\r\n\r\n2\r\n\r\n3\r\n\r\n4")]
+        public async Task PlainTextFunctionTest(string inputExp, string expectedResult)
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/PlainTextTests.cs:23]: Can these be merged into PlainText.txt?  EG, we may not need any dedicated C#  test.

Comment (CarlosFigueiraMSFT) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/PlainTextTests.cs:23]: I don't think the .txt tests support \r\n in the output. We may be able to work around it by using something like

    >> Substitute(Substitute(PlainText(\"1<  br  >2<  Li />3\")", "1\r\n2\r\n3"), Char(13), "\r"), Char(10), "\n")
    "1\r\n2\r\n3"

Or we may want to see if we can "fix" the test runner to account for multi-line strings

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 18 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:32]: Carlos mentioned use ignorecase for easy reading so I updated existing regex as well. Is it ok? or you want me remove ignorecase tage from RegExFlags?

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/PlainTextTests.cs (hunk 19 above)
Comment (nguhoa) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/PlainTextTests.cs]: Thanks for suggestion. Removed this file and add to test txt file.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 18 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs]: Updated RegExFlags for ignorecase here

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt [hunk 20]
```diff
@@ -0,0 +1,97 @@
+>> PlainText("")
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt:1]: >PlainText [](http://example.com/codeflow?start=3&length=9)

Just double checking: these are all correct for PowerApps ? This .txt will run against JS tests too.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 1 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:1076]: updated

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs (hunk 3 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs:27]: will remove this from PA after add to Fx.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 4 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:944]: Thanks

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 7 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs]: Updated here.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 9 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs]: updated.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 8 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs]: updated

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 21]
```diff
@@ -1059,6 +1071,34 @@ public static FormulaValue OptionSetValueToLogicalName(IRContext irContext, Opti
             var optionSet = args[0];
             var logicalName = optionSet.Option;
             return new StringValue(irContext, logicalName);
+        }
+
+        public static FormulaValue PlainText(IRContext irContext, StringValue[] args)
+        {        
+            //Decode html specific characters
+            string text = WebUtility.HtmlDecode(args[0].Value.Trim());
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:1079]: >string text = WebUtility.HtmlDecode(args[0].Value.Trim()); [](http://example.com/codeflow?start=12&length=58)

This may cause double decoding. For example, `PlainText("Hello&lt;br/&gt;world")` should return `"Hello<br/>world"` but it now adds a newline where it shouldn't.

Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs]: Move this to the end of function to not double decoding.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt (hunk 20 above)
Comment (nguhoa) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/PlainText.txt:1]: all these tests have the same results with PA.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by CarlosFigueiraMSFT — DISMISSED** [review]
🚢

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs (hunk 3 above)
Comment (anderson-joyle) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs:27]: You can use the following PR as a guide:
[https://msazure.visualstudio.com/OneAgile/_git/PowerApps-Client/pullrequest/7757449?_a=files](https://msazure.visualstudio.com/OneAgile/_git/PowerApps-Client/pullrequest/7757449?_a=files)

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 8 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs]: I agree with Carlos, we should inherit `StringOneArgFunction` and have this as `NoOpAlreadyHandledByIR` that way it can be handled at IR phase and the advantage is there other backend in future would not need to worry about the blank handling.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs (hunk 3 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/PlainText.cs:27]: Thanks, will use it.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 8 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1216]: Updated to use StringOneArgFunction.

**Review by jas-valgotar — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.
